tutorials/dataframe/tdf018_customActions.C

```diff
@@ -0,0 +1,77 @@
+/// \file
+/// \ingroup tutorial_tdataframe
+/// \notebook -draw
+/// This tutorial shows how to implement a custom action.
+/// As an example, we build a helper for filling THns.
+///
+/// \macro_code
+///
+/// \date April 2018
+/// \author Enrico Guiraud, Danilo Piparo
+
+template <typename T, unsigned int NDIM, typename... ColumnTypes>
+class THnHelper : public ROOT::Detail::TDF::TActionImpl<THnHelper<T, NDIM, ColumnTypes...>> {
+public:
+   // This is the list of the types of the columns
+   using ColumnTypes_t = ROOT::TypeTraits::TypeList<ColumnTypes...>;
+   using THn_t = THnT<T>;
+   using Result_t = THn_t;
+
+private:
+   std::vector<std::shared_ptr<THn_t>> fHistos; // one per slot
+   const ROOT::Detail::TDF::ColumnNames_t fColumnNames;
+
+public:
+   THnHelper(std::string_view name, std::string_view title, std::array<int, NDIM> nbins, std::array<double, NDIM> xmins,
+             std::array<double, NDIM> xmax, unsigned int nSlots, ROOT::Detail::TDF::ColumnNames_t columnNames)
+      : fHistos(nSlots, std::make_shared<THn_t>(std::string(name).c_str(), std::string(title).c_str(), NDIM,
```

Comment:
this creates a vector of shared_ptrs that all point to the same address (the one allocated by the make_shared call):

```c++
#include <cassert>
#include <memory>
#include <vector>

int main() {
   std::vector<std::shared_ptr<int>> ptrs(2, std::make_shared<int>(42));
   assert(ptrs[0] != ptrs[1]);
   return 0;
}
```

Reply:
ahaha, what a mistake! nice catch.